### PR TITLE
added support for bitbucket

### PR DESCRIPTION
**Pull request by AlessandroMinoccheri**
This should solve #337
I was thinking to create two of our services for Github and Bitbucket client to avoid so much code into strategy class, but I realized that this could be enough.
What do you think about this?

### CODE REVIEW

File: config/packages/bitbucket_api.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+services:
+  Bitbucket\Client:
+    class: Bitbucket\Client
+    calls:
+      - method: authenticate
+        arguments:
+          - '%env(BITBUCKET_AUTH_METHOD)%'
+          - '%env(BITBUCKET_TOKEN)%'
+          - '%env(BITBUCKET_SECRET)%'
```
Comment (JellyBellyDev) [inline config/packages/bitbucket_api.yaml:9]: please add new line at end of file.

File: phpunit.xml.dist [hunk 2]
```diff
@@ -39,6 +39,10 @@
         <!-- env name="TRUSTED_HOSTS" value="'^localhost|example\.com$'" -->
         <!-- ###- symfony/framework-bundle ### -->
 
+        <env name="BITBUCKET_AUTH_METHOD" value="http_password"/>
+        <env name="BITBUCKET_TOKEN" value="alessandro_minoccheri"/>
+        <env name="BITBUCKET_SECRET" value="pL4gSBDh2Bc4VZNvRrdm"/>
```
Comment (JellyBellyDev) [inline phpunit.xml.dist:44]: please don't publish your token here, but put this on `.env.test.local`! no?

File: src/Badge/Infrastructure/Package/PackageRepository.php [hunk 3]
```diff
@@ -48,26 +51,17 @@ public function fetchByRepository(string $repository): Package
 
         \preg_match('/(https)(:\/\/|@)([^\/:]+)[\/:]([^\/:]+)\/(.+)$/', $apiPackage->getRepository(), $matches);
 
-        if (isset($matches[4], $matches[5])) {
+        if ((isset($matches[4], $matches[5])) && ($apiPackage instanceof ApiPackage)) {
```
Comment (JellyBellyDev) [inline src/Badge/Infrastructure/Package/PackageRepository.php:54]: `isset` for `$matches[3]` no?

File: src/Badge/Service/ClientStrategy.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                ->repositories()
+                ->users($username)
+                ->show($repositoryName);
+
+            if (!$this->isValidBitbucketRepository($repoBitbucketData)) {
+                throw new RepositoryDataNotValid('Repository data not valid: '.(string) \json_encode($repoBitbucketData));
+            }
+
+            $defaultBranch = $repoBitbucketData['mainbranch']['name'];
+        }
```
Comment (JellyBellyDev) [inline src/Badge/Service/ClientStrategy.php:58]: switch case instead of double if?

File: tests/Badge/Service/ClientStrategyTest.php [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+        $apiInterface = $this->prophesize(Repo::class);
+        $apiInterface->show($this->username, $this->repositoryName)->willReturn(
+            [
+                'default_branch' => $defaultBranch,
+            ]
+        )->shouldBeCalledTimes(1);
+
+        $this->githubClient->api('repo')->willReturn($apiInterface)->shouldBeCalledTimes(1);
+        $source = 'github.com';
```
Comment (JellyBellyDev) [inline tests/Badge/Service/ClientStrategyTest.php:56]: +100 for not having used the const! ;)

**Comment by JellyBellyDev** [comment]
I found another interesting use case: https://packagist.org/packages/kiwa/template-basic
This should show the  badge: `composer.lock|committed` and `suggesters|2`
could you check please?

File: phpunit.xml.dist (hunk 2 above)
Comment (AlessandroMinoccheri) [reply phpunit.xml.dist:44]: Ouch!

**Comment by AlessandroMinoccheri** [comment]
@JellyBellyDev I have understood the problem, Github and Bitbucket have different URL for repositories, not only the host but the construction of all URL.
Asap I can fix the problem.

File: src/Badge/Service/ClientStrategy.php (hunk 4 above)
Comment (AlessandroMinoccheri) [reply src/Badge/Service/ClientStrategy.php:94]: Done

File: src/Badge/ValueObject/Repository.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    public function getUsername(): string
+    {
+        return $this->username;
+    }
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+}
```
Comment (JellyBellyDev) [inline src/Badge/ValueObject/Repository.php:54]: +1000

**Comment by JellyBellyDev** [comment]
LGTM 👍

**Pull request merged**